File: bank.java
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

class Transaction {
    private String transactionType;
    private double amount;
    private Date date;

    public Transaction(String transactionType, double amount, Date date) {
        this.transactionType = transactionType;
        this.amount = amount;
        this.date = date;
    }

    public String getTransactionType() {
        return transactionType;
    }

    public double getAmount() {
        return amount;
    }

    public Date getDate() {
        return date;
    }
}

class BankAccount {
    private int accountNumber;
    private String accountName;
    protected double balance;
    private List<Transaction> transactions;

    public BankAccount(int accountNumber, String accountName) {
        this.accountNumber = accountNumber;
        this.accountName = accountName;
        this.balance = 0;
        this.transactions = new ArrayList<>();
    }

    public void deposit(double amount) {
        if (amount > 0) {
            balance += amount;
            transactions.add(new Transaction("Deposit", amount, new Date()));
            System.out.println("Deposited " + amount + " into account " + accountNumber);
        } else {
            System.out.println("Invalid deposit amount");
        }
    }

    public void withdraw(double amount) {
        if (amount > 0 && amount <= balance) {
            balance -= amount;
            transactions.add(new Transaction("Withdrawal", amount, new Date()));
            System.out.println("Withdrew " + amount + " from account " + accountNumber);
        } else {
            System.out.println("Invalid withdrawal amount or insufficient balance");
        }
    }

    public void displayAccountDetails() {
        System.out.println("Account Number: " + accountNumber);
        System.out.println("Account Holder Name: " + accountName);
        System.out.println("Balance: " + balance);
    }

    public void displayTransactions() {
        System.out.println("Transaction History:");
        for (Transaction transaction : transactions) {
            System.out.println(transaction.getTransactionType() + ": " + transaction.getAmount() +
                    ", Date: " + transaction.getDate());
        }
    }
}

class SavingsAccount extends BankAccount {
    private double minBalance;

    public SavingsAccount(int accountNumber, String accountName, double minBalance) {
        super(accountNumber, accountName);
        this.minBalance = minBalance;
    }

    @Override
    public void withdraw(double amount) {
        if (balance - amount >= minBalance) {
            super.withdraw(amount);
        } else {
            System.out.println("Withdrawal failed: Insufficient balance or below minimum balance");
        }
    }

    @Override
    public void displayAccountDetails() {
        super.displayAccountDetails();
        System.out.println("Minimum Balance: " + minBalance);
    }
}

class RegisteredSavingsAccount extends SavingsAccount {
    private String registeredType;

    public RegisteredSavingsAccount(int accountNumber, String accountName, double minBalance, String registeredType) {
        super(accountNumber, accountName, minBalance);
        this.registeredType = registeredType;
    }

    @Override
    public void displayAccountDetails() {
        super.displayAccountDetails();
        System.out.println("Registered Type: " + registeredType);
    }
}

class CheckingAccount extends BankAccount {
    private double overdraftLimit;

    public CheckingAccount(int accountNumber, String accountName, double overdraftLimit) {
        super(accountNumber, accountName);
        this.overdraftLimit = overdraftLimit;
    }

    @Override
    public void withdraw(double amount) {
        if (balance - amount >= -overdraftLimit) {
            super.withdraw(amount);
        } else {
            System.out.println("Withdrawal failed: Insufficient funds and overdraft limit exceeded");
        }
    }

    @Override
    public void displayAccountDetails() {
        super.displayAccountDetails();
        System.out.println("Overdraft Limit: " + overdraftLimit);
    }
}

class Loan {
    private int accountNumber;
    private double loanAmount;
    private double remainingAmount;
    private double interestRate;

    public Loan(int accountNumber, double loanAmount, double interestRate) {
        this.accountNumber = accountNumber;
        this.loanAmount = loanAmount;
        this.remainingAmount = loanAmount;
        this.interestRate = interestRate;
    }

    public boolean makePayment(double amount) {
        if (amount > 0) {
            double interestAmount = remainingAmount * (interestRate / 100);
            double totalPayment = amount + interestAmount;
            if (totalPayment <= remainingAmount) {
                remainingAmount -= totalPayment;
                System.out.println("Payment of " + amount + " made for loan " + accountNumber);
                return true;
            } else {
                System.out.println("Payment exceeds remaining loan amount");
            }
        } else {
            System.out.println("Invalid payment amount");
        }
        return false;
    }

    public void displayLoanDetails() {
        System.out.println("Loan Account Number: " + accountNumber);
        System.out.println("Loan Amount: " + loanAmount);
        System.out.println("Remaining Amount: " + remainingAmount);
        System.out.println("Interest Rate: " + interestRate + "%");
    }
}

public class bank {
    public static void main(String[] args) {
        // Test the banking system
        BankAccount bankAccount = new BankAccount(1001, "Elon Musk");
        SavingsAccount savingsAccount = new SavingsAccount(2001, "Bill Gates", 100);
        RegisteredSavingsAccount registeredSavingsAccount = new RegisteredSavingsAccount(3001, "Steve Jobs", 500, "TFSA");
        CheckingAccount checkingAccount = new CheckingAccount(4001, "Jeff Bezos", 200);

        // Test transactions
        bankAccount.deposit(1000);
        bankAccount.withdraw(500);
        bankAccount.displayTransactions();
        System.out.println();

        savingsAccount.deposit(2000);
        savingsAccount.withdraw(150);
        savingsAccount.displayTransactions();
        System.out.println();

        registeredSavingsAccount.deposit(3000);
        registeredSavingsAccount.withdraw(400);  
        registeredSavingsAccount.displayTransactions();
        System.out.println();

        checkingAccount.deposit(3000);
        checkingAccount.withdraw(3500);  
        checkingAccount.displayAccountDetails();
        System.out.println();

        Loan loan = new Loan(5001, 10000, 5);
        loan.makePayment(1500);
        loan.displayLoanDetails();
    }
}

File: bank.py
from datetime import datetime

class Transaction:
    def __init__(self, transaction_type, amount, date):
        self.transaction_type = transaction_type
        self.amount = amount
        self.date = date


class BankAccount:
    def __init__(self, account_number, account_name, balance=0):
        self._account_number = account_number
        self._account_name = account_name
        self._balance = balance
        self._transactions = []

    def deposit(self, amount):
        if amount > 0:
            self._balance += amount
            self._transactions.append(Transaction("Deposit", amount, datetime.now()))
            print(f"Deposited {amount} into account {self._account_number}")
        else:
            print("Invalid deposit amount")

    def withdraw(self, amount):
        if amount > 0 and amount <= self._balance:
            self._balance -= amount
            self._transactions.append(Transaction("Withdrawal", amount, datetime.now()))
            print(f"Withdrew {amount} from account {self._account_number}")
        else:
            print("Invalid withdrawal amount or insufficient balance")

    def display_account_details(self):
        print(f"Account Number: {self._account_number}")
        print(f"Account Holder Name: {self._account_name}")
        print(f"Balance: {self._balance}")

    def display_transactions(self):
        print("Transaction History:")
        for transaction in self._transactions:
            print(f"{transaction.transaction_type}: {transaction.amount}, Date: {transaction.date}")


class SavingsAccount(BankAccount):
    def __init__(self, account_number, account_name, balance=0, min_balance=0):
        super().__init__(account_number, account_name, balance)
        self._min_balance = min_balance

    def withdraw(self, amount):
        if self._balance - amount >= self._min_balance:
            super().withdraw(amount)
        else:
            print("Withdrawal failed: Insufficient balance or below minimum balance")

    def display_account_details(self):
        super().display_account_details()
        print(f"Minimum Balance: {self._min_balance}")

class CheckingAccount(BankAccount):
    def __init__(self, account_number, account_name, balance=0, overdraft_limit=0):
        super().__init__(account_number, account_name, balance)
        self._overdraft_limit = overdraft_limit

    def withdraw(self, amount):
        if self._balance - amount >= -self._overdraft_limit:
            super().withdraw(amount)
        else:
            print("Withdrawal failed: Insufficient funds and overdraft limit exceeded")

    def display_account_details(self):
        super().display_account_details()
        print(f"Overdraft Limit: {self._overdraft_limit}")

class RegisteredSavingsAccount(SavingsAccount):
    def __init__(self, account_number, account_name, balance=0, min_balance=0, registered_type=''):
        super().__init__(account_number, account_name, balance, min_balance)
        self._registered_type = registered_type

    def display_account_details(self):
        super().display_account_details()
        print(f"Registered Type: {self._registered_type}")


class Loan:
    def __init__(self, account_number, loan_amount, interest_rate):
        self._account_number = account_number
        self._loan_amount = loan_amount
        self._interest_rate = interest_rate
        self._remaining_amount = loan_amount

    def make_payment(self, amount):
        if amount > 0:
            interest_amount = self._remaining_amount * (self._interest_rate / 100)
            total_payment = amount + interest_amount
            if total_payment <= self._remaining_amount:
                self._remaining_amount -= total_payment
                print(f"Payment of {amount} made for loan {self._account_number}")
                return True
            else:
                print("Payment exceeds remaining loan amount")
        else:
            print("Invalid payment amount")
        return False

    def display_loan_details(self):
        print(f"Loan Account Number: {self._account_number}")
        print(f"Loan Amount: {self._loan_amount}")
        print(f"Remaining Amount: {self._remaining_amount}")
        print(f"Interest Rate: {self._interest_rate}%")

def test_bank_accounts():
    # Create instances of BankAccount, SavingsAccount, and CheckingAccount
    bank_acct = BankAccount(1001, "Elon Musk")
    savings_acct = SavingsAccount(2001, "Bill Gates", min_balance=100)
    checking_acct = CheckingAccount(3001, "Steve Jobs", overdraft_limit=200)
    registered_savings_acct = RegisteredSavingsAccount(4001, "Jeff Bezos", min_balance=500, registered_type="TFSA")

    # Test transactions
    bank_acct.deposit(1000)
    bank_acct.withdraw(500)
    bank_acct.display_transactions()
    print()

    savings_acct.deposit(2000)
    savings_acct.withdraw(150)
    savings_acct.display_transactions()
    print()

    checking_acct.deposit(3000)
    checking_acct.withdraw(3500)  
    checking_acct.display_account_details()

    registered_savings_acct.deposit(3000)
    registered_savings_acct.withdraw(400)  
    registered_savings_acct.display_transactions()
    print()

    # Test loans
    loan = Loan(4001, 10000, 5)
    loan.make_payment(1500)
    loan.display_loan_details()


if __name__ == "__main__":
    test_bank_accounts()
